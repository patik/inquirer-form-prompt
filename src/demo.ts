import { Separator } from '@inquirer/core'
import { exit } from 'node:process'
import type { FormTheme } from './index.js'
import form from './index.js'

const errorHander = (error: unknown): void => {
    if (error instanceof Error && error.name === 'ExitPromptError') {
        console.log('✨')
        exit(0)
    } else {
        throw error
    }
}

process.on('uncaughtException', errorHander)

export const demo = async (
    {
        variant = 'table',
        dense = false,
    }: {
        variant: FormTheme['variant']
        dense: FormTheme['dense']
    } = { variant: 'table', dense: false },
): Promise<void> => {
    console.log()
    try {
        const answers = await form({
            message: 'Plan Your Adventure',
            submessage: 'Tell us about your dream trip!',
            fields: [
                new Separator('✈️ Trip Details'),
                {
                    label: 'Traveler Name',
                    type: 'text',
                    value: 'Firstname Lastname',
                    description: 'Your full name as it appears on your passport.',
                },
                {
                    label: 'Destination City',
                    type: 'text',
                    value: '',
                    description: 'Which city would you like to visit?',
                },
                new Separator('🏨 Preferences'),
                {
                    label: 'Include Hotels',
                    type: 'boolean',
                    value: true,
                    description: 'Should we include hotel recommendations in your itinerary?',
                },
                {
                    label: 'Train pass needed',
                    type: 'boolean',
                    value: false,
                    description: 'Will you need a train pass for your trip?',
                },
                new Separator('🚂 Transportation'),
                {
                    label: 'Preferred Transport',
                    type: 'radio',
                    choices: ['High-speed Train', 'Budget Flight', 'Scenic Bus Route'],
                    value: 'High-speed Train',
                    description: 'How would you like to travel between cities?',
                },
                {
                    label: 'Activities of Interest',
                    type: 'checkbox',
                    choices: ['Museums & Art', 'Local Cuisine', 'Historical Sites', 'Nightlife', 'Nature & Parks'],
                    value: ['Museums & Art', 'Local Cuisine'],
                    description: 'What activities interest you most? (Select all that apply)',
                },
            ],
            theme: {
                variant,
                dense,
            },
        })

        console.log('\n🎯 Your Travel Preferences:\n')
        console.log(
            answers
                .map((a) =>
                    a instanceof Separator
                        ? '─'.repeat(40)
                        : `${a.label}: ${
                              // It's preferred for the demo to print the value, warts and all
                              // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
                              a.value
                          }`,
                )
                .join('\n'),
        )
    } catch (error) {
        if (error instanceof Error && error.message === 'Escape key pressed') {
            console.log('\n✈️ Trip planning cancelled - maybe next time!')
        } else {
            throw error
        }
    }
}

// If the --run flag was used, run it immediately
// Read --variant and --dense flags from arguments
if (process.argv.includes('--run')) {
    await demo({
        variant:
            (process.argv.find((arg) => arg.startsWith('--variant='))?.slice(10) as FormTheme['variant']) || 'table',
        dense: process.argv.includes('--dense'),
    })
    exit(0)
}
