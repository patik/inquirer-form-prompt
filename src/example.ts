import { Separator } from '@inquirer/core'
import { form } from './index.js'

export const example = async (): Promise<void> => {
    console.log('')
    try {
        const answers = await form({
            message: 'Fill in the form',
            fields: [
                new Separator('Text fields'),
                {
                    name: 'Alpha',
                    type: 'text',
                    value: 'aaa aaa aaa',
                    description: 'This is a simple text field.',
                },
                {
                    name: 'Bravo',
                    type: 'text',
                    value: 'bbb',
                    description: 'Another simple text field.',
                },
                new Separator('Boolean fields'),
                {
                    name: 'Charlie',
                    type: 'boolean',
                    value: true,
                    description: 'This is a boolean field with a description.',
                },
                {
                    name: 'Delta',
                    type: 'boolean',
                    value: false,
                    description: 'This is another boolean field with a description.',
                },
                new Separator('Multiple choice'),
                {
                    name: 'Echo',
                    type: 'radio',
                    choices: ['foxtrot', 'golf', 'hotel'],
                    value: 'golf',
                    description: 'This is a radio field with a description.',
                },
                {
                    name: 'India',
                    type: 'checkbox',
                    choices: ['juliet', 'kilo', 'lima'],
                    value: ['juliet', 'lima'],
                    description: 'This is a checkbox field with a description.',
                },
            ],
        })

        console.log('\n\nForm values:\n')
        console.log(
            answers
                .map((a) =>
                    a instanceof Separator
                        ? '---------------'
                        : `${a.name}: ${
                              // It's preferred for the demo to print the value, warts and all
                              // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
                              a.value
                          }`,
                )
                .join('\n'),
        )
    } catch (error) {
        if (error instanceof Error && error.message === 'Escape key pressed') {
            console.log('\n‚ùå Form cancelled by user')
        } else {
            throw error
        }
    }
}
