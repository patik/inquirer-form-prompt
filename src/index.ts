import { createPrompt, isEnterKey, Separator, useKeypress, usePrefix, useState } from '@inquirer/core'
import ansiEscapes from 'ansi-escapes'
import Table from 'cli-table3'
import readline from 'readline'
import { bold, dim } from 'yoctocolors'
import { editBooleanField } from './keyHandlers/editBoolean.js'
import { editCheckboxField } from './keyHandlers/editCheckbox.js'
import { editRadioField } from './keyHandlers/editRadio.js'
import { editTextField } from './keyHandlers/editText.js'
import { handleNavigation } from './keyHandlers/handleNavigation.js'
import { fieldToTableRow } from './renderers/fieldToTableRow.js'
import type { Config, Fields, InternalFields, ReturnedItems } from './types.js'

function toInternalFields(fields: Fields): InternalFields {
    return fields.map((field) => {
        if (!(field instanceof Separator) && field.type === 'checkbox') {
            return {
                ...field,
                highlightIndex: 0,
            }
        }

        return field
    })
}

function getInitialIndex(fields: Fields): number {
    const firstNonSeparatorIndex = fields.findIndex((field) => !(field instanceof Separator))

    return firstNonSeparatorIndex >= 0 ? firstNonSeparatorIndex : 0
}

export const form = async (options: Config): Promise<ReturnedItems> => {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    })

    // @ts-expect-error Only way I know how to hide the cursor
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
    rl.output.write('\x1B[?25l')

    const answer = await createPrompt<ReturnedItems, Config>((config, done) => {
        const [fields, setFields] = useState<InternalFields>(toInternalFields(config.fields))
        const [selectedIndex, setSelectedIndex] = useState(() => getInitialIndex(config.fields))
        const prefix = usePrefix({})

        useKeypress((key, rl) => {
            if (isEnterKey(key)) {
                done(fields)
                return
            }

            if (handleNavigation({ fields, key, selectedIndex, setSelectedIndex, rl })) {
                return
            }

            if (key.name === 'escape') {
                // @ts-expect-error Only way I know how to signal that the user pressed escape
                done(Symbol('Escape key pressed'))
                return
            }

            const currentField = fields[selectedIndex]

            if (!currentField || currentField instanceof Separator) {
                return
            }

            if (currentField.type === 'text') {
                const nextFields = editTextField({ fields, currentField, key, selectedIndex, rl })
                setFields(nextFields)
                return
            }

            if (currentField.type === 'radio') {
                const nextFields = editRadioField({ fields, currentField, selectedIndex, key, rl })
                setFields(nextFields)
                return
            }

            if (currentField.type === 'checkbox') {
                const nextFields = editCheckboxField({ fields, currentField, key, selectedIndex, rl })
                setFields(nextFields)
                return
            }

            const nextFields = editBooleanField({ fields, currentField, selectedIndex, key, rl })

            setFields(nextFields)
            return
        })

        const message = config.message ? bold(config.message) : ''
        const submessage = config.submessage ? `\n\n${config.submessage}\n` : ''
        const tables: string[] = []
        let currentTable = new Table()
        let currentRows: Array<[string, string]> = []
        let tableFooter = ''

        fields.forEach((field, index) => {
            if (field instanceof Separator) {
                if (currentRows.length > 0) {
                    currentTable.push(...currentRows)
                    tables.push(currentTable.toString())
                    currentRows = []
                }

                tables.push(tableFooter)
                tables.push('')
                tables.push(field.separator)
                currentTable = new Table()
                tableFooter = ''

                return
            }

            const result = fieldToTableRow(selectedIndex)(field, index)
            if (!(result instanceof Separator)) {
                currentRows.push(result)
            }

            const isSelected = selectedIndex === index

            if (isSelected && !(field instanceof Separator) && field.description) {
                tableFooter = dim(`  ${field.description}`)
            }
        })

        // Push the last table if it has rows
        if (currentRows.length > 0) {
            currentTable.push(...currentRows)
            tables.push(currentTable.toString())
            tables.push(tableFooter)
        }

        return `${prefix} ${message}${submessage} ${dim('(tab/arrows to move between fields, enter to finish)')}

${tables.join('\n')}${ansiEscapes.cursorHide}
`
    })(options)

    // @ts-expect-error Only way I know how to show the cursor
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
    rl.output.write('\x1B[?25h')
    rl.close()

    // Check if the user aborted
    if (typeof answer === 'symbol' && String(answer) === 'Symbol(Escape key pressed)') {
        throw new Error('Escape key pressed')
    }

    return answer
}
