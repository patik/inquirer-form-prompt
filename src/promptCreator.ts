import type { Config, Fields, InquirerReadline, InternalFields, InternalFormField, ReturnedItems } from '@/util/types'
import type { KeypressEvent } from '@inquirer/core'
import { isEnterKey, Separator, useKeypress, usePrefix, useState } from '@inquirer/core'
import ansiEscapes from 'ansi-escapes'
import { bold, dim } from 'yoctocolors'
import { editBooleanField } from './keyHandlers/editBoolean.js'
import { editCheckboxField } from './keyHandlers/editCheckbox.js'
import { editRadioField } from './keyHandlers/editRadio.js'
import { editTextField } from './keyHandlers/editText.js'
import { handleNavigation } from './keyHandlers/handleNavigation.js'
import { toLabelTop } from './renderers/labelTop/toLabelTop.js'
import { toTable } from './renderers/table/toTable.js'

function toInternalFields(fields: Fields): InternalFields {
    return fields.map((field) => {
        if (!(field instanceof Separator) && field.type === 'checkbox') {
            return {
                ...field,
                highlightIndex: 0,
            }
        }

        return field
    })
}

function getInitialIndex(fields: Fields): number {
    const firstNonSeparatorIndex = fields.findIndex((field) => !(field instanceof Separator))

    return firstNonSeparatorIndex >= 0 ? firstNonSeparatorIndex : 0
}

const updateFields = ({
    fields,
    currentField,
    key,
    setFields,
    focusedIndex,
    rl,
}: {
    fields: InternalFields
    setFields: (newFields: InternalFields) => void
    currentField: InternalFormField
    key: KeypressEvent
    focusedIndex: number
    rl: InquirerReadline
}): void => {
    if (currentField.type === 'boolean') {
        const nextFields = editBooleanField({ fields, currentField, focusedIndex, key, rl })
        setFields(nextFields)
        return
    }

    if (currentField.type === 'radio') {
        const nextFields = editRadioField({ fields, currentField, focusedIndex, key, rl })
        setFields(nextFields)
        return
    }

    if (currentField.type === 'checkbox') {
        const nextFields = editCheckboxField({ fields, currentField, key, focusedIndex, rl })
        setFields(nextFields)
        return
    }

    const nextFields = editTextField({ fields, currentField, key, focusedIndex, rl })
    setFields(nextFields)
    return
}

/**
 * Exported for testing purposes
 */
export const promptCreator = (config: Config, done: (value: ReturnedItems) => void): string => {
    const [fields, setFields] = useState<InternalFields>(toInternalFields(config.fields))
    const [focusedIndex, setFocusedIndex] = useState(() => getInitialIndex(config.fields))
    const prefix = usePrefix({})

    useKeypress((key, rl) => {
        if (isEnterKey(key)) {
            done(fields)
            return
        }

        if (key.name === 'escape') {
            // @ts-expect-error This is the only way I know how to signal that the user pressed escape
            done(Symbol('Escape key pressed'))
            return
        }

        if (handleNavigation({ fields, key, focusedIndex, setFocusedIndex, rl })) {
            return
        }

        const currentField = fields[focusedIndex]

        if (!currentField || currentField instanceof Separator) {
            return
        }

        updateFields({
            fields,
            currentField,
            key,
            setFields,
            focusedIndex,
            rl,
        })
    })

    const message = config.message ? bold(config.message) : ''
    const submessage = config.submessage ? `\n\n${config.submessage}\n` : ''
    const fieldOutput =
        config.theme?.variant === 'label-top'
            ? toLabelTop(fields, focusedIndex, config.theme.dense)
            : toTable(fields, focusedIndex)

    return `${prefix} ${message}${submessage} ${dim('(tab/arrows to move between fields, enter to finish)')}
${fieldOutput}${ansiEscapes.cursorHide}
`
}
