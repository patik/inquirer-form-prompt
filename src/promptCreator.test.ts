import { createPrompt, Separator } from '@inquirer/core'
import { render } from '@inquirer/testing'
import { promptCreator } from 'src/promptCreator'
import type { Config, ReturnedItems } from 'src/util/types'
import { describe, expect, it } from 'vitest'

// Create a proper prompt function compatible with @inquirer/testing
const formPrompt = createPrompt<ReturnedItems, Config>(promptCreator)

describe('Form Prompt', () => {
    describe('Basic Rendering', () => {
        it('should render a form with text fields', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'Please fill out this form',
                fields: [
                    {
                        type: 'text',
                        label: 'Full Name',
                        value: 'John Doe',
                    },
                    {
                        type: 'text',
                        label: 'Email',
                        value: 'john@example.com',
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('Please fill out this form')
            expect(screen).toContain(' → Full Name')
            expect(screen).toContain('  Email')
            expect(screen).toContain('John Doe')
            expect(screen).toContain('john@example.com')
            expect(screen).toContain('(tab/arrows to move between fields, enter to finish)')
        })

        it('should render a form with boolean fields', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'Settings',
                fields: [
                    {
                        type: 'boolean',
                        label: 'Enable notifications',
                        value: true,
                    },
                    {
                        type: 'boolean',
                        label: 'Auto-save',
                        value: false,
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('? Settings')
            expect(screen).toContain(' → Enable notifications')
            expect(screen).toContain('  Auto-save')
            expect(screen).toContain('true')
            expect(screen).toContain('false')
        })

        it('should render a form with radio fields', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'Select option',
                fields: [
                    {
                        type: 'radio',
                        label: 'Preferred language',
                        choices: ['JavaScript', 'TypeScript', 'Python'],
                        value: 'TypeScript',
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('Select option')
            expect(screen).toContain('Preferred language')
            expect(screen).toContain('TypeScript')
        })

        it('should render a form with checkbox fields', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'Select technologies',
                fields: [
                    {
                        type: 'checkbox',
                        label: 'Technologies',
                        choices: ['React', 'Vue', 'Angular', 'Svelte'],
                        value: ['React', 'Vue'],
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('Select technologies')
            expect(screen).toContain('Technologies')
            expect(screen).toContain('☒ React  ☒ Vue  ☐ Angular  ☐ Svelte')
        })

        it('should render a form with separators', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'User Profile',
                fields: [
                    {
                        type: 'text',
                        label: 'Name',
                        value: 'John',
                    },
                    new Separator('--- Contact Information ---'),
                    {
                        type: 'text',
                        label: 'Email',
                        value: 'john@example.com',
                    },
                    {
                        type: 'text',
                        label: 'Phone',
                        value: '+1234567890',
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('User Profile')
            expect(screen).toContain('Name')
            expect(screen).toContain('--- Contact Information ---')
            expect(screen).toContain('Email')
            expect(screen).toContain('Phone')
        })

        it('should render submessage when provided', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'Main message',
                submessage: 'Additional instructions here',
                fields: [
                    {
                        type: 'text',
                        label: 'Field',
                        value: '',
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('Main message')
            expect(screen).toContain('Additional instructions here')
        })

        it('should work without a main message', async () => {
            const { getScreen } = await render(formPrompt, {
                fields: [
                    {
                        type: 'text',
                        label: 'Field',
                        value: 'value',
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('Field')
            expect(screen).toContain('value')
            expect(screen).toContain('(tab/arrows to move between fields, enter to finish)')
        })
    })

    describe('Field Selection and Cursor', () => {
        it('should handle initial field selection correctly', async () => {
            const { getScreen } = await render(formPrompt, {
                fields: [
                    new Separator('--- Header ---'),
                    {
                        type: 'text',
                        label: 'First selectable field',
                        value: '',
                    },
                    {
                        type: 'text',
                        label: 'Second field',
                        value: '',
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('First selectable field')
            expect(screen).toContain('Second field')
            // The first selectable field should be focused (indicated by arrow)
            expect(screen).toContain('→')
        })

        it('should show field description when field is focused', async () => {
            const { getScreen } = await render(formPrompt, {
                fields: [
                    {
                        type: 'text',
                        label: 'API Key',
                        value: '',
                        description: 'Enter your API key from the dashboard',
                    },
                    {
                        type: 'boolean',
                        label: 'Debug mode',
                        value: false,
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('API Key')
            expect(screen).toContain('Enter your API key from the dashboard')
        })
    })

    describe('Navigation and Interaction', () => {
        it('should allow navigation between fields with arrow keys', async () => {
            const { getScreen, events } = await render(formPrompt, {
                message: 'Test Navigation',
                fields: [
                    {
                        type: 'text',
                        label: 'First Field',
                        value: 'value1',
                    },
                    {
                        type: 'text',
                        label: 'Second Field',
                        value: 'value2',
                    },
                    {
                        type: 'boolean',
                        label: 'Third Field',
                        value: true,
                    },
                ],
            })

            // Initially, first field should be focused
            expect(getScreen()).toContain('→ First Field')

            // Navigate down to second field
            events.keypress('down')
            expect(getScreen()).toContain('→ Second Field')

            // Navigate down to third field
            events.keypress('down')
            expect(getScreen()).toContain('→ Third Field')

            // Navigate back up
            events.keypress('up')
            expect(getScreen()).toContain('→ Second Field')
        })

        it('should complete the form when enter is pressed', async () => {
            const { answer, events } = await render(formPrompt, {
                message: 'Complete this form',
                fields: [
                    {
                        type: 'text',
                        label: 'Username',
                        value: 'testuser',
                    },
                    {
                        type: 'boolean',
                        label: 'Active',
                        value: true,
                    },
                ],
            })

            // Press enter to submit the form
            events.keypress('enter')

            await expect(answer).resolves.toEqual([
                {
                    type: 'text',
                    label: 'Username',
                    value: 'testuser',
                },
                {
                    type: 'boolean',
                    label: 'Active',
                    value: true,
                },
            ])
        })

        it('should handle escape key to cancel', async () => {
            const { answer, events } = await render(formPrompt, {
                message: 'Test Cancel',
                fields: [
                    {
                        type: 'text',
                        label: 'Field',
                        value: 'value',
                    },
                ],
            })

            // Press escape to cancel
            events.keypress('escape')

            expect(await answer).toBeTypeOf('symbol')
            // eslint-disable-next-line @typescript-eslint/no-base-to-string
            expect((await answer).toString()).toMatch('Escape key pressed')
        })
    })

    describe('Mixed Field Types', () => {
        it('should render complex forms with all field types', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'Complete Survey',
                submessage: 'Please fill out all fields',
                fields: [
                    {
                        type: 'text',
                        label: 'Full Name',
                        value: 'Jane Doe',
                    },
                    {
                        type: 'text',
                        label: 'Email',
                        value: 'jane@example.com',
                    },
                    new Separator('--- Preferences ---'),
                    {
                        type: 'boolean',
                        label: 'Subscribe to newsletter',
                        value: true,
                    },
                    {
                        type: 'radio',
                        label: 'Preferred contact method',
                        choices: ['Email', 'Phone', 'SMS'],
                        value: 'Email',
                    },
                    {
                        type: 'checkbox',
                        label: 'Topics of interest',
                        choices: ['Technology', 'Sports', 'Music', 'Travel'],
                        value: ['Technology', 'Music'],
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('Complete Survey')
            expect(screen).toContain('Please fill out all fields')
            expect(screen).toContain('Full Name')
            expect(screen).toContain('Email')
            expect(screen).toContain('--- Preferences ---')
            expect(screen).toContain('Subscribe to newsletter')
            expect(screen).toContain('Preferred contact method')
            expect(screen).toContain('Topics of interest')
        })
    })

    describe('Edge Cases', () => {
        it('should handle fields with undefined values', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'Edge cases',
                fields: [
                    {
                        type: 'text',
                        label: 'Optional field',
                        value: undefined,
                    },
                    {
                        type: 'boolean',
                        label: 'Optional boolean',
                        value: undefined,
                    },
                    {
                        type: 'radio',
                        label: 'Optional radio',
                        choices: ['A', 'B'],
                        value: undefined,
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('Edge cases')
            expect(screen).toContain('Optional field')
            expect(screen).toContain('Optional boolean')
            expect(screen).toContain('Optional radio')
        })

        it('should handle empty arrays for checkbox choices', async () => {
            const { getScreen } = await render(formPrompt, {
                fields: [
                    {
                        type: 'checkbox',
                        label: 'Empty choices',
                        choices: [],
                        value: [],
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('Empty choices')
        })

        it('should handle very long field names', async () => {
            const { getScreen } = await render(formPrompt, {
                fields: [
                    {
                        type: 'text',
                        label: 'This is a very long field name that should be handled properly by the form renderer without breaking the layout',
                        value: 'test value',
                    },
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('This is a very long field name')
            expect(screen).toContain('test value')
        })

        it('should handle forms with only separators', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'Only separators',
                fields: [
                    new Separator('--- Section 1 ---'),
                    new Separator('--- Section 2 ---'),
                    new Separator('--- Section 3 ---'),
                ],
            })

            const screen = getScreen()
            expect(screen).toContain('Only separators')
            expect(screen).toContain('--- Section 1 ---')
            expect(screen).toContain('--- Section 2 ---')
            expect(screen).toContain('--- Section 3 ---')
        })

        it('should handle empty fields array', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'Empty form',
                fields: [],
            })

            const screen = getScreen()
            expect(screen).toContain('Empty form')
            expect(screen).toContain('(tab/arrows to move between fields, enter to finish)')
        })
    })

    describe('Snapshots', () => {
        it('should match snapshot for basic form', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'User Information',
                fields: [
                    {
                        type: 'text',
                        label: 'First Name',
                        value: 'John',
                        description: 'Enter your first name',
                    },
                    {
                        type: 'text',
                        label: 'Last Name',
                        value: 'Doe',
                    },
                    {
                        type: 'boolean',
                        label: 'Subscribe',
                        value: true,
                    },
                ],
            })

            expect(getScreen()).toMatchSnapshot()
        })

        it('should match snapshot for complex form with separators', async () => {
            const { getScreen } = await render(formPrompt, {
                message: 'Complete Profile',
                submessage: 'Fill out all sections',
                fields: [
                    new Separator('Personal Information'),
                    {
                        type: 'text',
                        label: 'Name',
                        value: 'Jane Smith',
                    },
                    {
                        type: 'text',
                        label: 'Email',
                        value: 'jane@example.com',
                    },
                    new Separator('Preferences'),
                    {
                        type: 'radio',
                        label: 'Theme',
                        choices: ['light', 'dark', 'auto'],
                        value: 'dark',
                    },
                    {
                        type: 'checkbox',
                        label: 'Languages',
                        choices: ['JavaScript', 'TypeScript', 'Python'],
                        value: ['JavaScript', 'TypeScript'],
                    },
                ],
            })

            expect(getScreen()).toMatchSnapshot()
        })
    })
})
