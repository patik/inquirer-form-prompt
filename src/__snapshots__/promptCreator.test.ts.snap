// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formPrompt > Snapshots > should match snapshot for basic form 1`] = `
"? User Information (tab/arrows to move between fields, enter to finish)
┌──────────────┬───────────────────┐
│ → First Name │ John              │
├──────────────┼───────────────────┤
│   Last Name  │ Doe               │
├──────────────┼───────────────────┤
│   Subscribe  │  ◉ true  ◯ false  │
└──────────────┴───────────────────┘
  Enter your first name"
`;

exports[`formPrompt > Snapshots > should match snapshot for complex form with separators 1`] = `
"? Complete Profile

Fill out all sections
 (tab/arrows to move between fields, enter to finish)


Personal Information
┌─────────┬──────────────────┐
│ → Name  │ Jane Smith       │
├─────────┼──────────────────┤
│   Email │ jane@example.com │
└─────────┴──────────────────┘


Preferences
┌─────────────┬────────────────────────────────────────┐
│   Theme     │  ◯ light  ◉ dark  ◯ auto               │
├─────────────┼────────────────────────────────────────┤
│   Languages │  ☒ JavaScript  ☒ TypeScript  ☐ Python  │
└─────────────┴────────────────────────────────────────┘"
`;
